src/mqt/bench/targets/gatesets/ionq.py

```diff
@@ -0,0 +1,283 @@
+    """
+
+    def __init__(
+        self,
+        phi0: ParameterValueType,
+        phi1: ParameterValueType,
+        theta: ParameterValueType | None = 0.25,
+        label: str | None = None,
+    ) -> None:
+        """Create new MS gate."""
+        super().__init__(
+            "ms",
+            2,
+            [phi0, phi1, theta],
+            label=label,
+        )
+
+    def _define(self) -> None:
+        """Define the MS gate."""
+        phi0 = self.params[0]
+        phi1 = self.params[1]
+        theta = self.params[2]
+        q = QuantumRegister(2, "q")
+        alpha = phi0 + phi1
+        beta = phi0 - phi1
+
+        qc = QuantumCircuit(q)
+        qc.cx(q[1], q[0])
+        qc.x(q[0])
+        qc.cu(
+            2 * theta * np.pi,
+            2 * alpha * np.pi - np.pi / 2,
+            np.pi / 2 - 2 * alpha * np.pi,
+            0,  # gamma
+            control_qubit=q[0],
+            target_qubit=q[1],
+        )
+        qc.x(q[0])
+        qc.cu(
+            2 * theta * np.pi,
+            -2 * beta * np.pi - np.pi / 2,
+            np.pi / 2 + 2 * beta * np.pi,
+            0,  # gamma
+            control_qubit=q[0],
+            target_qubit=q[1],
+        )
+        qc.cx(q[1], q[0])
+
+        self.definition = qc
+
+
+class ZZGate(Gate):  # type: ignore[misc]
+    r"""Two-qubit ZZ-rotation gate.
+
+    **Circuit Symbol:**
+    .. parsed-literal::
+        q_0: ───■────
+                │zz(θ)
+        q_1: ───■────
+    **Matrix Representation:**.
+
+    .. math::
+
+        ZZ(\theta) =
+            \begin{pmatrix}
+                e^{-i \theta*\pi} & 0 & 0 & 0 \\
+                0 & e^{i \theta*\pi} & 0 & 0 \\
+                0 & 0 & e^{i \theta*\pi} & 0 \\
+                0 & 0 & 0 & e^{-i \theta\*\pi}
+            \end{pmatrix}
+    """
+
+    def __init__(self, theta: ParameterValueType, label: str | None = None) -> None:
+        """Create new ZZ gate."""
+        super().__init__("zz", 2, [theta], label=label)
+
+    def _define(self) -> None:
+        """Define the ZZ gate."""
+        theta = self.params[0]
+        q = QuantumRegister(2, "q")
+        qc = QuantumCircuit(q)
+        qc.rzz(2 * np.pi * theta, 0, 1)
+
+        self.definition = qc
+
+
+def u_gate_equivalence() -> None:
+    """Add U gate equivalence to the SessionEquivalenceLibrary."""
+    q = QuantumRegister(1, "q")
+    theta_param = Parameter("theta_param")
+    phi_param = Parameter("phi_param")
+    lambda_param = Parameter("lambda_param")
+    u_gate = QuantumCircuit(q)
+    # this sequence can be compacted if virtual-z gates will be introduced
+    u_gate.append(GPI2Gate(0.5 - lambda_param / (2 * np.pi)), [0])
+    u_gate.append(
+        GPIGate(theta_param / (4 * np.pi) + phi_param / (4 * np.pi) - lambda_param / (4 * np.pi)),
+        [0],
+    )
+    u_gate.append(GPI2Gate(0.5 + phi_param / (2 * np.pi)), [0])
+    SessionEquivalenceLibrary.add_equivalence(UGate(theta_param, phi_param, lambda_param), u_gate)
```

Comment:
I believe it would be better if these methods would take an equivalence library as a parameter instead of unconditionally adding to the SEL.
Gives a bit more flexibility.

Reply:
I followed your hint. Additionally, I add the library now only immediately before the `transpile` method is called. I played around a bit with the `EquivalenceLibrary` and you can get all available decompositions for a gate but would need to compare the existing ones with the one that might be added to check if it is already in there - which is also not ideal.